Reworks gym equipment

JohnFulpWillard opened the pull request:
## About The Pull Request

Gym equipment are currently ported from Goon code, and I didn't like how it currently worked as it was buggy and the code was a mess. I decided to just rework how they worked entirely. I left the parts that weren't Goon code (like tooltips & deconstruction) alone because it's not goon code and I think it's fine that way. 

- Now you buckle yourself to it (like a chair, I didn't like how it didn't work like one) and can work out with an action button you get on buckle, you can do as many sets as you want, then unbuckle & go.
- You aren't hardstunned for 8 seconds for clicking on it anymore, you aren't flying around because the stun doesn't even last the duration of the animation anymore.

Video demonstration

https://github.com/tgstation/tgstation/assets/53777086/f5ae86f1-65fb-46c6-8a5f-8f9d9f0548b3

## Why It's Good For The Game

Reworks old code
Removes un-documented Goon code
Makes gym equipment less broken than it currently is.

## Changelog

:cl:
qol: Gym equipment was reworked. You now buckle yourself to weight machines to use them, rather than clicking on it and getting stunned for a few seconds. It also means it works like a chair now.
qol: Gym equipment no longer breaks a sweat in no gravity.
/:cl:

JohnFulpWillard committed:
Reworks gym equipment

Gym equipment are currently ported from Goon code, and I didn't like how it currently worked as it was buggy and the code was a mess. I decided to just rework how they worked entirely.
Now you buckle yourself to it and can work out with an action button you get on buckle, you can do as many sets as you want, then unbuckle & go.
You aren't hardstunned for 8 seconds for clicking on it anymore, you aren't flying around because the stun doesn't even last the duration of the animation anymore.

The animation of the weight machines and your character are off sync but it's also something dependent on process now so I didn't want to bother trying to sync it locally.

JohnFulpWillard commented on code/game/objects/structures/gym/weight_machine_action.dm:
```diff
@@ -0,0 +1,22 @@
+ * The action button given by the weight machine's buckle.
+ * This allows users to manually trigger working out.
+ */
+/datum/action/push_weights
+	name = "Work out"
+	desc = "Start working out"
```
I wrote this while on 24 hours no sleep so let me know any suggestions for it

JohnFulpWillard commented:
i forgot the maps

JohnFulpWillard committed:
Updatepaths & Maps

I removed a subtype of weight machine, so this is the updatepaths.

Sealed101 commented:
i liked how people were using gym equipment as a barrier for rooms like tech storage, was a curious way to timegate access to areas

Likteer commented:
@Sealed101 :)

Now about the PR.
The lack of sync is horrendous.

And what are we trading it away for?
Doing multiple sets.
Which there is no reason to do.
Why

atlasle commented:
yeah i'm with likteer, this is worse than outright removing them

Loafin34 commented:
sync it, and keep the stun after either cancelling the process or finishing it, as you need to catch your breath after such a hard workout. remove stun if there's no gravity, as it wasn't hard to pull/push weights.

Pumpkinoe commented:
is that bloodmoon rising???

JohnFulpWillard commented:
> keep the stun after either cancelling the process or finishing it

The stun was just keeping you in place as you worked out, not once you were done.

Also turns out the weight machine isn't actually in-sync on master either. I don't know why I assumed that it was.

> is that bloodmoon rising???

sang by the 4 peters

JohnFulpWillard committed:
Sync up animation, bonus only if gravity, & stun

JohnFulpWillard commented:
Added a new stun when you finish working out so I can negate it when there's no gravity.
Fixed the plane to be above the mob, and synced it up better than even pre-this pr


https://github.com/tgstation/tgstation/assets/53777086/fe1d52a3-f746-47db-8db5-fc20c971dfa6

JohnFulpWillard committed:
2 seconds of stun instead

san7890 commented on code/game/objects/structures/gym/punching_bag.dm:
```diff
@@ -0,0 +1,69 @@
+/obj/structure/punching_bag/Initialize(mapload)
+	. = ..()
+
+	AddElement( \
+		/datum/element/contextual_screentip_bare_hands, \
+		lmb_text = "Punch", \
+	)
```
does this really have to be multilined? if it needs it it's probably fine but it really doesn't appear to need it at-a-glance

JohnFulpWillard replied:
this is copy pasted from when i split the .dm files.

san7890 commented on code/game/objects/structures/gym/punching_bag.dm:
```diff
@@ -0,0 +1,69 @@
+
+/obj/structure/punching_bag/wrench_act_secondary(mob/living/user, obj/item/tool)
+	tool.play_tool_sound(src)
+	if(anchored)
+		balloon_alert(user, "unsecured")
+		anchored = FALSE
+	else
+		balloon_alert(user, "secured")
+		anchored = TRUE
```
we can make this way more simple/agnostic with the following:

```suggestion
	balloon_alert(user, anchored ? "unsecured" : "secured")
	anchored = !anchored
```

if you think it's ugly and don't want it - it's fine, but there is a better option here.

san7890 commented on code/game/objects/structures/gym/punching_bag.dm:
```diff
@@ -0,0 +1,69 @@
+
+/obj/structure/punching_bag/crowbar_act_secondary(mob/living/user, obj/item/tool)
+	if(anchored)
+		balloon_alert(user, "unsecure first!")
```
```suggestion
		balloon_alert(user, "still secured!")
```

san7890 commented on code/game/objects/structures/gym/punching_bag.dm:
```diff
@@ -0,0 +1,69 @@
+		return FALSE
+	tool.play_tool_sound(src)
+	balloon_alert(user, "deconstructing...")
+	if (!do_after(user, 10 SECONDS, target = src))
+		return FALSE
```
```suggestion
	if (!do_after(user, 10 SECONDS, target = src))
		balloon_alert(user, "cancelled!")
		return FALSE
		
```
i really like having a cancelled message if the do_after fails, although you can probably ignore this as well

JohnFulpWillard replied:
It is obvious that it's been cancelled from the fact you are no longer near it and the do after bar is gone, it doesn't need a balloon alert

san7890 replied:
do after bars have been historically flakey for me but whatever assuming the optimal case

san7890 commented on code/game/objects/structures/gym/weight_machine.dm:
```diff
@@ -0,0 +1,131 @@
+
+/obj/structure/weightmachine/wrench_act_secondary(mob/living/user, obj/item/tool)
+	tool.play_tool_sound(src)
+	if(anchored)
+		balloon_alert(user, "unsecured")
+		anchored = FALSE
+	else
+		balloon_alert(user, "secured")
+		anchored = TRUE
```
ditto on making it more concise

san7890 commented on code/game/objects/structures/gym/weight_machine.dm:
```diff
@@ -0,0 +1,131 @@
+	return TRUE
+
+/obj/structure/weightmachine/crowbar_act_secondary(mob/living/user, obj/item/tool)
+	if(anchored)
+		balloon_alert(user, "unsecure first!")
+		return FALSE
+	tool.play_tool_sound(src)
+	balloon_alert(user, "deconstructing...")
+	if (!do_after(user, 10 SECONDS, target = src))
+		return FALSE
+	new /obj/item/stack/sheet/iron(get_turf(src), 5)
+	new /obj/item/stack/rods(get_turf(src), 2)
+	new /obj/item/chair(get_turf(src))
+	qdel(src)
+	return TRUE
```
ditto on those balloon alerts

san7890 commented on code/game/objects/structures/gym/weight_machine.dm:
```diff
@@ -0,0 +1,131 @@
+	var/datum/action/push_weights/weight_action
+
+	///List of messages picked when using the machine.
+	var/static/list/more_weight = list(
+		"pushing it to the limit!",
+		"going into overdrive!",
+		"burning with determination!",
+		"rising up to the challenge!",
+		"getting strong now!",
+		"getting ripped!",
+	)
```
ehhhh balloon alerts shouldn't really have exclamation points when they aren't failure states but i think these are cool enough to remain, just pointing it out since i saw it

JohnFulpWillard replied:
I think it's flavorful since it is kinda a big deal

san7890 reviewed (changes requested):
half of these are optional

JohnFulpWillard committed:
Update code/game/objects/structures/gym/punching_bag.dm

Co-authored-by: san7890 <the@san7890.com>

JohnFulpWillard committed:
Update weight_machine.dm

JohnFulpWillard committed:
Merge branch 'GYM-EQUIPMENT' of https://github.com/JohnFulpWillard/tgstation into GYM-EQUIPMENT

JohnFulpWillard committed:
still secured
